File: app-service-comments/src/main/java/com/github/arhor/aws/graphql/federation/comments/service/impl/UserServiceImpl.java
package com.github.arhor.aws.graphql.federation.comments.service.impl;

import com.github.arhor.aws.graphql.federation.comments.data.entity.UserEntity;
import com.github.arhor.aws.graphql.federation.comments.data.repository.UserRepository;
import com.github.arhor.aws.graphql.federation.comments.generated.graphql.DgsConstants.USER;
import com.github.arhor.aws.graphql.federation.comments.generated.graphql.types.User;
import com.github.arhor.aws.graphql.federation.comments.service.UserService;
import com.github.arhor.aws.graphql.federation.comments.util.Caches;
import com.github.arhor.aws.graphql.federation.common.exception.EntityNotFoundException;
import com.github.arhor.aws.graphql.federation.common.exception.Operation;
import com.github.arhor.aws.graphql.federation.tracing.Trace;
import jakarta.annotation.PostConstruct;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.Cache;
import org.springframework.cache.CacheManager;
import org.springframework.stereotype.Service;

import java.util.UUID;

import static com.github.arhor.aws.graphql.federation.comments.util.CacheManagerUtils.getCache;

@Trace
@Service
@RequiredArgsConstructor
public class UserServiceImpl implements UserService {

    private final CacheManager cacheManager;
    private final UserRepository userRepository;

    private Cache cache;

    @PostConstruct
    public void initialize() {
        cache = getCache(cacheManager, Caches.IDEMPOTENT_ID_SET);
    }

    @Override
    public User findInternalUserRepresentation(final UUID userId) {
        return userRepository.findById(userId)
            .map(this::mapEntityToUser)
            .orElseThrow(() -> new EntityNotFoundException(
                USER.TYPE_NAME,
                USER.Id + " = " + userId,
                Operation.LOOKUP
            ));
    }

    @Override
    public void createInternalUserRepresentation(final UUID userId, final UUID idempotencyKey) {
        cache.get(idempotencyKey, () ->
            userRepository.save(
                new UserEntity(userId)
            )
        );
    }

    @Override
    public void deleteInternalUserRepresentation(final UUID userId, final UUID idempotencyKey) {
        cache.get(idempotencyKey, () -> {
            userRepository.deleteById(userId);
            return null;
        });
    }

    private User mapEntityToUser(final UserEntity entity) {
        return User.newBuilder()
            .id(entity.id())
            .build();
    }
}

File: app-service-posts/src/main/kotlin/com/github/arhor/aws/graphql/federation/posts/service/impl/UserServiceImpl.kt
package com.github.arhor.aws.graphql.federation.posts.service.impl

import com.github.arhor.aws.graphql.federation.common.exception.EntityNotFoundException
import com.github.arhor.aws.graphql.federation.common.exception.Operation
import com.github.arhor.aws.graphql.federation.posts.data.entity.UserEntity
import com.github.arhor.aws.graphql.federation.posts.data.repository.UserRepository
import com.github.arhor.aws.graphql.federation.posts.generated.graphql.DgsConstants
import com.github.arhor.aws.graphql.federation.posts.generated.graphql.types.User
import com.github.arhor.aws.graphql.federation.posts.service.UserService
import com.github.arhor.aws.graphql.federation.posts.util.Caches
import com.github.arhor.aws.graphql.federation.posts.util.get
import com.github.arhor.aws.graphql.federation.tracing.Trace
import jakarta.annotation.PostConstruct
import org.springframework.cache.Cache
import org.springframework.cache.CacheManager
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import java.util.UUID

@Trace
@Service
class UserServiceImpl(
    private val cacheManager: CacheManager,
    private val userRepository: UserRepository,
) : UserService {

    private lateinit var cache: Cache

    @PostConstruct
    fun initialize() {
        cache = cacheManager[Caches.IDEMPOTENT_ID_SET]
    }

    override fun findInternalUserRepresentation(userId: UUID): User {
        return userRepository.findByIdOrNull(userId)?.let(::mapEntityToUser)
            ?: throw EntityNotFoundException(
                DgsConstants.USER.TYPE_NAME,
                "${DgsConstants.USER.Id} = $userId",
                Operation.LOOKUP,
            )
    }

    override fun createInternalUserRepresentation(userId: UUID, idempotencyId: UUID) {
        cache.get(idempotencyId) {
            userRepository.save(UserEntity(id = userId))
        }
    }

    override fun deleteInternalUserRepresentation(userId: UUID, idempotencyId: UUID) {
        cache.get(idempotencyId) {
            userRepository.deleteById(userId)
        }
    }

    private fun mapEntityToUser(entity: UserEntity): User {
        return User(
            id = entity.id,
        )
    }
}
